lib/cache/cache_test.go

```diff
@@ -384,11 +385,211 @@ func waitForEvent(c *check.C, eventsC <-chan Event, expectedEvent string, skipEv
+		c.Assert(p.trustS.UpsertCertAuthority(ca), check.IsNil)
+	}
+
+	for i := 0; i < inits; i++ {
+		var err error
+
+		p.cacheBackend, err = memory.New(
+			memory.Config{
+				Context: context.TODO(),
+				Mirror:  true,
+			})
+		c.Assert(err, check.IsNil)
+
+		// simulate bad connection to auth server
+		p.backend.SetReadError(trace.ConnectionProblem(nil, "backend is unavailable"))
+		p.eventsS.closeWatchers()
+
+		p.cache, err = New(ForAuth(Config{
+			Context:       context.TODO(),
+			Backend:       p.cacheBackend,
+			Events:        p.eventsS,
+			ClusterConfig: p.clusterConfigS,
+			Provisioner:   p.provisionerS,
+			Trust:         p.trustS,
+			Users:         p.usersS,
+			Access:        p.accessS,
+			DynamicAccess: p.dynamicAccessS,
+			Presence:      p.presenceS,
+			AppSession:    p.appSessionS,
+			RetryPeriod:   200 * time.Millisecond,
+			EventsC:       p.eventsC,
+			PreferRecent: PreferRecent{
+				Enabled: true,
+			},
+		}))
+		c.Assert(err, check.IsNil)
+
+		p.backend.SetReadError(nil)
+
+		cas, err := p.cache.GetCertAuthorities(services.UserCA, false)
+		// we don't actually care whether the cache ever fully constructed
+		// the CA list.  for the purposes of this test, we just care that it
+		// doesn't return the CA list *unless* it was successfully constructed.
+		if err == nil {
+			c.Assert(len(cas), check.Equals, caCount)
+		} else {
+			fixtures.ExpectConnectionProblem(c, err)
+		}
+
+		c.Assert(p.cache.Close(), check.IsNil)
+		p.cache = nil
+		c.Assert(p.cacheBackend.Close(), check.IsNil)
+		p.cacheBackend = nil
+	}
+}
+
+// TestCompletenessReset verifies that flaky backends don't cause
+// the cache to return partial results during reset.
+func (s *CacheSuite) TestCompletenessReset(c *check.C) {
+	const caCount = 100
+	const resets = 20
+	p := s.newPackWithoutCache(c, ForAuth)
+	defer p.Close()
+
+	// put lots of CAs in the backend
+	for i := 0; i < caCount; i++ {
+		ca := suite.NewTestCA(services.UserCA, fmt.Sprintf("%d.example.com", i))
+		c.Assert(p.trustS.UpsertCertAuthority(ca), check.IsNil)
+	}
+
+	var err error
+	p.cache, err = New(ForAuth(Config{
+		Context:       context.TODO(),
+		Backend:       p.cacheBackend,
+		Events:        p.eventsS,
+		ClusterConfig: p.clusterConfigS,
+		Provisioner:   p.provisionerS,
+		Trust:         p.trustS,
+		Users:         p.usersS,
+		Access:        p.accessS,
+		DynamicAccess: p.dynamicAccessS,
+		Presence:      p.presenceS,
+		AppSession:    p.appSessionS,
+		RetryPeriod:   200 * time.Millisecond,
+		EventsC:       p.eventsC,
+		PreferRecent: PreferRecent{
+			Enabled: true,
+		},
+	}))
+	c.Assert(err, check.IsNil)
+
+	// verify that CAs are immediately available
+	cas, err := p.cache.GetCertAuthorities(services.UserCA, false)
+	c.Assert(err, check.IsNil)
+	c.Assert(len(cas), check.Equals, caCount)
+
+	for i := 0; i < resets; i++ {
+		// simulate bad connection to auth server
+		p.backend.SetReadError(trace.ConnectionProblem(nil, "backend is unavailable"))
```

Comment:
can you help me understand the effect of setting the backend error followed by resetting it to nil? What does it mean for cache? Is it supposed to affect background activity and if so - in what way?

Reply:
This backed simulates the backed of the auth server.  Temporarily putting it into an error state simulates losing connectivity with the auth server (or some similar problem which causes all connectivity to the upstream state to be lost).

Reply:
I understand that, but it looks like it might be too short of a time window for error to be seen - and I wanted to make sure I understand how the error injection is supposed to work.